File: 2024/Day 6/main.py

from multiprocessing import Process, Value
with open("input.txt") as f:
    data = [list(line) for line in f.read().split("\n")]

guard_pos = None
MAX_X, MAX_Y = len(data[0]), len(data)
obstacles, possible_spots = [], []
for y, row in enumerate(data):
    for x, cell in enumerate(row):
        if cell in ("^", "v", "<", ">"): guard_pos = ((x, y), cell)
        elif cell == "#": obstacles.append((x, y))
        else: possible_spots.append((x, y))

def get_next_pos(pt, direction):
    x, y = pt
    if direction == "^": return (x, y - 1)
    if direction == "v": return (x, y + 1)
    if direction == "<": return (x - 1, y)
    if direction == ">": return (x + 1, y)

def get_new_direction(direction):
    if direction == "^": return ">"
    if direction == "v": return "<"
    if direction == "<": return "^"
    if direction == ">": return "v"

def pointInGrid(pt):
    x, y = pt
    return 0 <= x < MAX_X and 0 <= y < MAX_Y

def simulate(guard_pos, obstacles):
    path = {guard_pos}
    duplicates = 0
    next_pos = get_next_pos(*guard_pos)
    while (pointInGrid(next_pos)) and duplicates < len(path):
        if next_pos in obstacles:
            new_direction = get_new_direction(guard_pos[1])
            next_pos = get_next_pos(guard_pos[0], new_direction)
            while next_pos in obstacles:
                new_direction = get_new_direction(new_direction)
                next_pos = get_next_pos(guard_pos[0], new_direction)
            guard_pos = (next_pos, new_direction)
        else:
            guard_pos = (next_pos, guard_pos[1])
        if guard_pos[0] in path: duplicates += 1
        else: 
            path.add(guard_pos[0])
        next_pos = get_next_pos(*guard_pos)
    if duplicates >= len(path): return -1, path
    return len(path), path

value, path = simulate(guard_pos, obstacles)
print(f"Part 1: {value}")

N = len(possible_spots)
N_per_process = N // 15
possible_spots = list(filter(lambda x: x in path, possible_spots))
lsts = [possible_spots[i:i + N_per_process] for i in range(0, N, N_per_process)]
count = Value('i', 0)

def worker(lst, obstacles):
    for spot in lst:
        obstacles_copy = obstacles.copy()
        obstacles_copy.append(spot)
        if simulate(guard_pos, obstacles_copy)[0] == -1: count.value += 1

processes = []
for i, lst in enumerate(lsts, 1):
    p = Process(target=worker, args=(lst, obstacles))
    p.start()
    processes.append(p)
    
for p in processes:
    p.join()
    
print(f"Part 2: {count.value}")

File: 2024/Day 6/main.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <vector>
#include <array>
#include <set>
#include <algorithm>
#include <thread>
#include <atomic>
#define MAX 500

using namespace std;

typedef struct Point {
    int x;
    int y;

    Point() {}
    Point(int x, int y) : x(x), y(y) {}

    bool operator==(const Point& other) const {
        return x == other.x && y == other.y;
    }
    bool operator<(const Point& other) const {
        return x < other.x || (x == other.x && y < other.y);
    }
} Point;

typedef struct Position {
    Point point;
    char direction;

    Position() {}
    Position(Point point, char direction) : point(point), direction(direction) {}
} Position;

vector<Point> grid;
vector<Point> obstacles;
vector<array<Point, MAX>> possible_spots;
Position guard;
int MAX_X, MAX_Y;

void readInput() {
    ifstream file("input.txt");
    string line;
    int y = 0;
    array<Point, MAX> spots;
    int i = 0;
    while (std::getline(file, line)) {
        MAX_X = line.size();
        for (int x = 0; x < line.size(); x++) {
            grid.emplace_back(x, y);

            if (line[x] == '^' || line[x] == 'v' || line[x] == '<' || line[x] == '>') {
                guard = Position{{x, y}, line[x]};
            } else if (line[x] == '#') {
                obstacles.emplace_back(x, y); 
            } else {
                if (i < MAX) {
                    spots[i] = {x, y};
                    i++;
                } else {
                    possible_spots.push_back(spots);
                    i = 0;
                    spots[i] = {x, y};
                    i++;
                }
            }     
        }
        y++;
    }
    possible_spots.push_back(spots);     
    MAX_Y = y;
}

Point getNextPoint(Position position) {
    Point point = position.point;
    char direction = position.direction;
    switch (direction) {
        case '^':
            return {point.x, point.y - 1};
        case '>':
            return {point.x + 1, point.y};
        case 'v':
            return {point.x, point.y + 1};
        case '<':
            return {point.x - 1, point.y};
        default:
            return point;
    }
}

char getNextDirection(char direction) {
    switch (direction) {
        case '^':
            return '>';
        case '>':
            return 'v';
        case 'v':
            return '<';
        case '<':
            return '^';
        default:
            return direction;
    }
}

bool pointInGrid(Point point) {
    return point.x >= 0 && point.x < MAX_X && point.y >= 0 && point.y < MAX_Y;
}

int simulate(Position guard, vector<Point>& obstacles) {
    set<Point> visited = {guard.point};
    int duplicates = 0;
    Point nextPoint = getNextPoint(guard);
    while (pointInGrid(nextPoint) && duplicates < visited.size()) {
        if (find(obstacles.begin(), obstacles.end(), nextPoint) != obstacles.end()) {
            char nextDirection = getNextDirection(guard.direction);
            nextPoint = getNextPoint({guard.point, nextDirection});
            while (pointInGrid(nextPoint) && find(obstacles.begin(), obstacles.end(), nextPoint) != obstacles.end()) {
                nextDirection = getNextDirection(nextDirection);
                nextPoint = getNextPoint({guard.point, nextDirection});
            }
            guard = {nextPoint, nextDirection};
        } else {
            guard.point = nextPoint;
        }
        if (visited.find(guard.point) != visited.end()) duplicates++;
        else visited.insert(guard.point);
        nextPoint = getNextPoint(guard);
    }
    if (duplicates >= visited.size()) return -1;
    return visited.size();
}

void worker(const array<Point, MAX>& lst, const vector<Point>& obstacles, atomic<int>& count) {
    for (Point spot : lst) {
        vector<Point> new_obstacles = obstacles;
        new_obstacles.push_back(spot);
        int result = simulate(guard, new_obstacles);
        if (result == -1) count++;
    }
}

int main() {
    readInput();
    printf("Part 1: %d\n", simulate(guard, obstacles));
    atomic<int> count(0);
    vector<thread> threads;
    
    for (int i = 0; i < possible_spots.size(); ++i) {
        threads.emplace_back(worker, cref(possible_spots[i]), cref(obstacles), ref(count));
    }

    for (thread& thread : threads) {
        thread.join();
    }
    printf("Part 2: %d\n", count.load());
    return 0;
}